antelope_contracts/contracts/erc20/src/erc20.cpp

```diff
@@ -189,48 +196,61 @@ void erc20::onbridgemsg(const bridge_message_t &message) {
 
     check(msg.data.size() >= 4, "not enough data in bridge_message_v0");
-    if (*(uint32_t *)&(msg.data[0]) == 0xe5323365 /* big-endian 0x653332e5 */) {
-        //sha("bridgeTransferV0(address,uint,string)") = 0xe5323365
-        check(msg.data.size() >= 4 + kAddressLength + 32, "not enough data in bridge_message_v0 of application type 0xe5323365");
+
+    uint32_t app_type = 0;
+    memcpy((void *)&app_type, (const void *)&(msg.data[0]), sizeof(app_type));
+
+    if (app_type == 0xe5323365 /* big-endian 0x653332e5 */) {
+        // abi.encodeWithSignature("bridgeTransferV0(address,uint,string)", to, amount, memo);
+        // sha("bridgeTransferV0(address,uint,string)") = 0x653332e5
+        check(msg.data.size() >= 4 + 32 /*to*/ + 32 /*amount*/ + 32 /*memo offset*/ + 32 /*memo len*/, 
+            "not enough data in bridge_message_v0 of application type 0x653332e5");
 
         evmc::address dest_addr;
-        memcpy(dest_addr.bytes, (void *)&(msg.data[4]), kAddressLength);
+        memcpy(dest_addr.bytes, (void *)&(msg.data[4 + 32 - kAddressLength]), kAddressLength);
         std::optional<uint64_t> dest_acc = silkworm::extract_reserved_address(dest_addr);
         check(!!dest_acc, "destination address in bridge_message_v0 must be reserved address");
 
-        uint8_t amount_bytes[32]={};
-        memcpy(amount_bytes, (void *)&(msg.data[4 + kAddressLength]), 32);
-        
-        intx::uint256 value{}, zero{0_u256};
-        value = intx::be::load<intx::uint256>(amount_bytes);
+        auto read_uint256 = [&](const auto &msg, size_t offset) -> intx::uint256 {
+            uint8_t buffer[32]={};
+            check(msg.data.size() >= offset + 32, "not enough data in bridge_message_v0 of application type 0x653332e5");
+            memcpy(buffer, (void *)&(msg.data[offset]), 32);
+            return intx::be::load<intx::uint256>(buffer);
+        };
 
-        intx::uint256 mult = intx::exp(10_u256, intx::uint256(itr->erc20_precision - itr->min_deposit.symbol.precision()));
-        check(value % mult == zero, "bridge amount can not have dust");
+        intx::uint256 value = read_uint256(msg, 4 + 32);
+        intx::uint256 mult = intx::exp(10_u256, intx::uint256(itr->erc20_precision - itr->ingress_fee.symbol.precision()));
+        check(value % mult == 0_u256, "bridge amount can not have dust");
         value /= mult;
 
         uint64_t dest_amount = (uint64_t)value;
         check(intx::uint256(dest_amount) == value && dest_amount < (1ull<<62)-1, "bridge amount value overflow");
         check(dest_amount > 0, "bridge amount must be positive");
 
+        check(read_uint256(msg, 4 + 32 + 32) == 96_u256, "invalid memo offset in bridge_message_v0");
+
+        intx::uint256 memo_len_ = read_uint256(msg, 4 + 32 + 32 + 32);
+        size_t memo_len = (uint64_t)(memo_len_);
+        check(memo_len_ <= 256_u256 && msg.data.size() >= 4 + 32 + 32 + 32 + 32 + memo_len, 
```

Comment:
not really and it depends on the memo length. in fact, msg len = 4/*func hash*/ + 32 /*to*/ + 32 /*amount*/ + 32 /*memo offset*/ + 32 /*memo len*/ + memo_data len align to 32 bytes

Reply:
So if we wanted to make that `==` it would require doing `+ ((memo_len + 31)/32)*32` (basically rounded up to nearest multiple of 32) rather than just `+ memo_len`.

But it sounds like we don't really care if the caller puts garbage data at the end anyway. So I guess we can keep it as is.